engine/renderer/private/particles/particle_pipeline.cpp

```diff
@@ -87,92 +125,159 @@ void ParticlePipeline::RecordCommands(vk::CommandBuffer commandBuffer, uint32_t
+    commandBuffer.bindIndexBuffer(indexBuffer, 0, vk::IndexType::eUint32);
+
+    CulledInstances* culledData = reinterpret_cast<CulledInstances*>(culledIndicesBuffer->mappedPtr);
+    commandBuffer.drawIndexed(6, culledData->count, 0, 0, 0, vkContext->Dldi());
 
+    commandBuffer.endRenderingKHR(vkContext->Dldi());
     util::EndLabel(commandBuffer, vkContext->Dldi());
 }
 
-void ParticlePipeline::UpdateEmitters(std::shared_ptr<ECS> ecs)
+void ParticlePipeline::UpdateEmitters(vk::CommandBuffer commandBuffer)
 {
-    auto view = ecs->registry.view<EmitterComponent>();
+    auto vkContext { _context->VulkanContext() };
+    auto resources { _context->Resources() };
+
+    auto view = _ecs->registry.view<EmitterComponent>();
     for (auto entity : view)
     {
         auto& component = view.get<EmitterComponent>(entity);
         if (component.timesToEmit != 0)
         {
             // TODO: do something with particle type later
+            component.emitter.randomValue = static_cast<float>(rand()) / static_cast<float>(RAND_MAX);
             _emitters.emplace_back(component.emitter);
             spdlog::info("Emitter received!");
             component.timesToEmit--; // TODO: possibly move the updating of emitters to the GPU
         }
     }
-}
-
-void ParticlePipeline::UpdateBuffers(vk::CommandBuffer commandBuffer)
-{
-    auto vkContext { _context->VulkanContext() };
 
-    // TODO: check if this swapping works
-    std::swap(_particlesBuffers[static_cast<uint32_t>(ParticleBufferUsage::eAliveNew)], _particlesBuffers[static_cast<uint32_t>(ParticleBufferUsage::eAliveCurrent)]);
-
-    if (_emitters.size() > 0)
+    if (!_emitters.empty())
     {
         vk::DeviceSize bufferSize = _emitters.size() * sizeof(Emitter);
 
         vmaCopyMemoryToAllocation(vkContext->MemoryAllocator(), _emitters.data(), _stagingBufferAllocation, 0, bufferSize);
-        util::CopyBuffer(commandBuffer, _stagingBuffer, _context->Resources()->BufferResourceManager().Access(_emittersBuffer)->buffer, bufferSize);
-
-        vk::BufferMemoryBarrier barrier {};
-        barrier.buffer = _context->Resources()->BufferResourceManager().Access(_emittersBuffer)->buffer;
-        barrier.size = bufferSize;
-        barrier.offset = 0;
-        barrier.srcAccessMask = vk::AccessFlagBits::eTransferWrite;
-        barrier.dstAccessMask = vk::AccessFlagBits::eShaderRead;
-
-        commandBuffer.pipelineBarrier(vk::PipelineStageFlagBits::eTransfer, vk::PipelineStageFlagBits::eComputeShader, vk::DependencyFlags { 0 }, {}, barrier, {});
+        util::CopyBuffer(commandBuffer, _stagingBuffer, resources->BufferResourceManager().Access(_emittersBuffer)->buffer, bufferSize);
     }
 }
 
+void ParticlePipeline::UpdateAliveLists()
+{
+    std::swap(_particlesBuffers[static_cast<uint32_t>(ParticleBufferUsage::eAliveNew)], _particlesBuffers[static_cast<uint32_t>(ParticleBufferUsage::eAliveCurrent)]);
+    UpdateParticleBuffersDescriptorSets();
```

Comment:
I would still make a task or improvement for making this data use a bindless model. That would remove all this descriptor bloat and make it simpler for swapping, since you can pass the indices as a push constant